libs/shaders/cringine/shaders/shader_program_builder.cpp

```diff
@@ -0,0 +1,116 @@
+#include "shader_program_builder.hpp"
+
+#include <cringine/utils/load_from_file.hpp>
+
+#include <GL/glew.h>
+
+#include <iostream>
+#include <utility>
+#include <array>
+
+using namespace cringine::shaders;
+
+namespace
+{
+    void log_shader_error(GLuint shader, GLuint shader_type)
+    {
+        const short array_size = 512;
+        std::array<GLchar, array_size> info_log{};
+        std::string shader_type_name;
+        glGetShaderInfoLog(shader, array_size, nullptr, info_log.data());
+        switch (shader_type) {
+            case GL_VERTEX_SHADER:
+                shader_type_name = "VERTEX";
+                break;
+            case GL_FRAGMENT_SHADER:
+                shader_type_name = "FRAGMENT";
+                break;
+            case GL_GEOMETRY_SHADER:
+                shader_type_name = "GEOMETRY";
+                break;
+                [[unlikely]] default : shader_type_name = "UNKNOWN";
+        }
+        std::cout << "ERROR::SHADER::" << shader_type_name << "::COMPILATION_FAILED\n"
+                  << info_log.data() << std::endl;
+    }
+
+    void log_shader_program_error(GLuint shader_program)
+    {
+        const short array_size = 512;
+        std::array<GLchar, array_size> info_log{};
+        glGetProgramInfoLog(shader_program, array_size, nullptr, info_log.data());
+        std::cout << "ERROR::SHADER_PROGRAM::LINK_FAILED\n"
+                  << info_log.data() << std::endl;
+    }
+
+    std::pair<std::vector<const char*>, std::vector<int>> to_char_array(const std::vector<std::string>& data)
+    {
+        std::vector<const char*> result;
+        std::vector<int> result_lengths;
+        result.reserve(data.size());
+        result_lengths.reserve(data.size());
+        for (const auto& str : data) {
+            result.push_back(str.c_str());
+            result_lengths.push_back(static_cast<int>(str.size()));
+        }
+        return {std::move(result), std::move(result_lengths)};
+    }
+
+    GLuint compile_shader(const std::vector<std::string>& shader_source, GLuint shader_type)
+    {
+        auto [shader_data, shader_data_lengths] = to_char_array(shader_source);
+
+        GLuint shader = glCreateShader(shader_type);
+        glShaderSource(shader, shader_data.size(), shader_data.data(), shader_data_lengths.data());
+        glCompileShader(shader);
+
+        GLint success{};
+        glGetShaderiv(shader, GL_COMPILE_STATUS, &success);
+
+        if (success == GL_FALSE) {
+            log_shader_error(shader, shader_type);
+        }
+
+        return shader;
+    }
+
+    GLuint create_shader_program(const std::vector<std::string>& vertex_shader_data, const std::vector<std::string>& fragment_shader_data)
```

Comment:
Think about using string_views instead basic strings as it might decrease memory usage

Reply:
And how do you suggest I convert the `std::vector<std:: string>` in `std::vector<std::string_view>`? Do you think it would be more optimal to allocate a new array than to pass it by reference?

Reply:
I suggest to use string_views at the stage where you create initial vector of strings

Reply:
But I'm reading it from file
I do not understand. Explain, please

Reply:
You could read file into string and after that move it to string_view. But I'm not actually sure how beneficial it's gonna be

Reply:
And as far as I aware of using references is not thread-safe so usage of string_view could eliminate that problem

Reply:
Can we suspend it for later? You can write a note in the project board

Reply:
Of course